runtime/runtime.m

```diff
@@ -1929,12 +1929,77 @@ -(void) xamarinSetGCHandle: (int) gc_handle;
 	} else {
 		/* If we're a wrapper type, we need to unregister here, since we won't enter the release trampoline */
 		xamarin_unregister_nsobject (self, managed_obj, &exception_gchandle);
-		// This lock is needed so that we can safely call retainCount in the toggleref callback.
+
+		//
+		// This lock is needed so that we can safely call retainCount in the
+		// toggleref callback.
+		//
+		// The race is between the following actions (given a managed object Z):
+		//
+		//   a1) This thread nulls out the handle for Z
+		//   a2) This thread calls release on Z's original handle
+		//   b1) Another thread fetches the handle for Z
+		//   b2) Another thread calls retainCount on Z's handle
+		//
+		// Possible execution orders:
+		//
+		//   1) a1-*: all such orders are safe, because b1 will read NULL and
+		//      b2 won't do anything
+		//   2) b1-b2-a1-a2: retainCount before release is safe.
+		//   3) b1-a1-b2-a2: retainCount before release is safe.
+		//   4) b1-a1-a2-b2: unsafe; this tries to call retainCount after
+		//      release.
+		//
+		// Order 4 would look like this:
+		//
+		//   * Thread A runs a GC, and starts calling toggleref callbacks.
+		//   * Thread A fetches the handle (H) for object Z in a toggleref
```

Comment:
Minor: it's _slightly_ confusing that you're flipping thread A and B in this description and below. "Thread A fetches the handle (H) for object Z in a toggleref" is b1 above.